Mayana Fitri

mayanafit opened the pull request:
ENV:
PORT=3000
SECRET=VUEVUE

admin:
email = admin@email.com
password = 12345

link: 
server = https://murmuring-headland-34171.herokuapp.com
client = https://e-store-334f4.web.app

mayanafit committed:
Server done based on TDD
+TDD for User, Banner, and Product all passed
+.gitignore is universal
+DOC REST API server done

nadia12 commented on server/controllers/UserController.js:
```diff
@@ -0,0 +1,49 @@
+
+        User.findOne({where: {email}})
+        .then(result => {
+            if (!result) {
```
secara logic, sudah oke, namun, coba sekarang ketika ngoding, latihan belajar menghindari nested if (if dalam if). 

ex: 

```
if ( !result || (result && comparePass(password, result.password)) ) {
     throw error
} else {
    let {id, name, email, role} = result
    let access_token = encode({id, name, email, role})
    res.status(200).json({access_token})
}
```

sample aja, ini silakan dicoba dulu,

mayanafit replied:
siap kakk, done perubahannya

nadia12 commented on server/helpers/jwt.js:
```diff
@@ -0,0 +1,12 @@
+const jwt = require('jsonwebtoken')
+const secret = process.env.SECRET
+
+const encode = (data) => {
```
[OPTIONAL]

FYI untuk arrow function, bila memang hanya 1 line, bisa langsung:
```
const encode = (data) =>  jwt.sign(data, secret)
```

mayanafit replied:
siapp makasih kak sudah kubenarkan hehehe

nadia12 commented on server/middlewares/auth.js:
```diff
@@ -0,0 +1,45 @@
+    let userData = decode(access_token)
+    req.user = userData
+    User.findByPk(userData.id)
+    .then(result => {
```
btw sudah dicoba kah, sekiranya user not found, kehandle kah errornya? `if(!result)` ?

mayanafit replied:
sudah kakk, karna access_tokennya jatohnya gaada/invalid masuk ke error 401

nadia12 commented on server/middlewares/auth.js:
```diff
@@ -0,0 +1,45 @@
+        statusCode: 403,
+        message: `You don't have access to this.`
+    }
+
```
kan di authentication, kamu sudah decode `req.user = userData`,  kalau sedari awal, saat ` jwt sign`, kamu menambahkan role,, ini bisa langsung kondisional berdasar req.user.role.. gak perlu query findByPk lagi..

mayanafit replied:
done kak, kemarin sempat bingung karena yang di .set adalah access_tokennya, lupa kalo ada req.user hehehe

nadia12 commented on server/middlewares/errHandler.js:
```diff
@@ -0,0 +1,27 @@
+        statusCode = 400
+        message = `Email already registered, please use another email.`
+    } else if (err.name === `SequelizeValidationError`) {
+        let errors = err.errors.map(e => e.message)
```
kenapa gak langsung `message = err.errors.map(e => e.message)` ? ><

mayanafit replied:
wkwk oiya lupa kak, suda kuganti juga

nadia12 commented on server/routes/product.js:
```diff
@@ -0,0 +1,12 @@
+const router = require(`express`).Router()
+const ProductController = require(`../controllers/ProductController`)
+const {authorization} = require(`../middlewares/auth`)
+
```
kenapa gak pakai `router.use('authorization')` ? tapi pilihan sih

mayanafit replied:
sudah ku ganti kak, hehehe kemarin mau make sure biar lulus tes sebenernya :')

mayanafit committed:
Correction for server code
-jwt function shortened
-change the authorization position at routes/index.js only for banner and product routes
-error message for error 400 condition shortened at middlewares/errHandler.js
-authorization logic changed, using req.user data at middlewares/auth.js
-no longer using nested if in login function at controllers/UserController.js

mayanafit committed:
Additional Validation
+add validation for role attribute  at models/user.js
+renew validation testing for role attribute at test/user.test.js
+already created db and registering admin as user for e-commerce CMS client

mayanafit committed:
Client side:
+create vue-cli under client folder
+add vue component in views folder:
 -BannerTable.vue
 -Dashboard.vue
 -FormBanner.vue
 -FormProduct.vue
 -Login.vue
 -NotFound.vue
 -ProductTable.vue
 -StatusStore.vue (still empty)
+add vue component in components folder:
 -TableDataBanner.vue
 -TableDataProduct.vue
+routing configuration for views component done
+guard navigation before and after login condition done
+all API process are in client/src/store.js
+CRUD operation for Products and Banners done

mayanafit committed:
Server side:
+add order option (sorting) in both Product.findAll and Banner.findAll controller at folder controllers

mayanafit committed:
Client side:
+table title in each product and banner page

mayanafit committed:
change title in index.html

mayanafit committed:
heroku configuration

mayanafit committed:
firebase configuration

mayanafit committed:
Client side:
+change width of name in Banner Table at src/component/TableDataBanner.vue
+change price format in Product Table at src/component/TableDataProduct.vue
+change route for edit/add form for routing Banner and Product

mayanafit committed:
Client side:
+add button cancel in every form, will direct to each table (product/banner)
+add relation between banner and product (banner has many products)
+add column 'set banner' in product table page
